pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Producer.java

```diff
@@ -132,6 +132,21 @@ public interface Producer<T> extends Closeable {
      * @since 2.7.0
      */
     TypedMessageBuilder<T> newMessage(Transaction txn);
+
+    /**
+     * Create a new message builder with transaction and schema, not required same parameterized type with the
+     * producer.
+     *
+     * <p>After the transaction commit, it will be made visible to consumer.
+     *
+     * <p>After the transaction abort, it will never be visible to consumer.
+     *
+     * @return a typed message builder that can be used to construct the message to be sent through this producer
+     * @see #newMessage()
+     */
+    <V> TypedMessageBuilder<V> newMessage(Schema<V> schema,
+                                          Transaction txn);
```

Comment:
I have a concern about backward compatibility with this interface change. Adding a new method without a default implementation will break source compatibility for any custom implementations of the `Producer` interface.
While it might be uncommon for users to implement this interface directly, it could happen in testing scenarios or custom proxy implementations. Java source compatibility rules require all implementors to provide all interface methods.

We could address the source compatibility issue by adding a default implementation:

```suggestion
    default <V> TypedMessageBuilder<V> newMessage(Schema<V> schema, Transaction txn) {
        // to retain source compatibilty of Producer interface for implementors of the Producer interface
        throw new UnsupportedOperationException(
                String.format("Method not implemented in: %s", getClass().getName()));
    }
```

The benefit of making it source compatible is that we could also backport this change to the 4.0.x client without introducing breaking changes.

Reply:
Do we really need to backport this change to 4.0.x client? LTS releases should not introduce such new APIs even if it does not break compatibility.

Reply:
Once it's ported to 4.0.x, the forward compatibility will be broken. According to https://github.com/apache/pulsar/issues/15966, we should also guarantee the downgrade compatibility

Reply:
Resolve this conversation. We can open another thread for the backport topic. For a new feature, adding a new non-default method should be accepted

Reply:
> Do we really need to backport this change to 4.0.x client? LTS releases should not introduce such new APIs even if it does not break compatibility.

This is a feature which would benefit existing 4.0.x users. By making the solution compatible without introducing breaking changes to the interface, we could deliver this feature to 4.0.x users immediately. Obviously there's the question whether we need to do this or not.

Reply:
I reviewed the PIP-175 and LTS releases recently again. PIP-175 mainly talked about the server side compatibility. Regarding the client side,

> Note that this consideration is separate from the compatibility between clients and brokers, where we never break compatibility. The oldest available Pulsar client can still talk with the newest Pulsar broker, and vice versa, a new client, will be perfectly fine with an older broker (except the new features won't be working).

From the existing definitions, it's acceptable to cherry-pick this **client-side** feature for 4.0.x. It does not break any compatibility: with this new API, the latest 4.0.x client can still access an older broker (that supports transaction).

The only question is the source compatibility you mentioned by making it `default`. Could you show a concrete example for how could a non-default new method affect users?